### PR TITLE
New features & minor patches for CST-decoding application

### PR DESCRIPTION
**Pull request by JRBCH**
Collection of features, convenience functions & minor bug fixes as part of the publication of recurrent SNNs for decoding of cortical spike trains.

These additions are necessary to run the networks implemented in https://github.com/fmi-basel/neural-decoding-RSNN.

**New features:**
- Support for learning rate schedulers
- Support for half-precision models

**Minor patches:**
- `RasDataset` now supports custom dtypes (needed for half precision models)
- `RasDataset` correctly passes the `data_augmentation` argument on
- `data_augmentation` not hardcoded as `True` in `SpikingDataset` anymore
- `InputGroup` passes a `kwargs` dict to `CellGroup` (this enables dropout for input layers)
- `StandardGenerator` now passes on a `worker_init_fn` argument that can be used for reproducible multi-process data loading by providing a random seed (see https://pytorch.org/docs/stable/notes/randomness.html)
- `nb_time_steps` can be independently set without re-configuring the model now

### CODE REVIEW

File: stork/datasets.py
Comment (jgygax) [inline stork/datasets.py:1]: Looks good to me. 
We might have another look in the dtypes of the RasDataset, as a binary input would be cool to at least be int...

File: stork/models.py [hunk 1]
```diff
@@ -315,6 +365,13 @@ def get_metrics_string(self, metrics_array, prefix="", postfix=""):
             s = s + " %s=%.3g" % (name, val)
         return s
 
+    def get_metrics_dict(self, metrics_array, prefix="", postfix=""):
```
Comment (jgygax) [inline stork/models.py:368]: Looks good, but the metric dict functions could maybe be put into one at some point

**Review by jgygax — APPROVED** [review]
The changes look good to me, thanks for putting these together here

File: stork/datasets.py
Comment (JRBCH) [reply stork/datasets.py:1]: Data storage is already in integer (neuron, time) tuples, so that is great for sparse input rasters.
For feeding the input, the problem is that the dtype has to match the dtype of neuronal states for torch op's to work correctly. So int8 works if the neuron model is quantized accordingly.

File: stork/models.py (hunk 1 above)
Comment (JRBCH) [reply stork/models.py:368]: agreed. TBH I am not sure why we need so many functions for the metrics. 
For a future re-factor, we should think hard about whether the metrics should be tracked by the LossStack class at all or whether we can should make a metric tracker
